sql/analyzer/join_order_builder.go

```diff
@@ -159,28 +159,69 @@ func (j *joinOrderBuilder) reorderJoin(n sql.Node) {
 // populateSubgraph recursively tracks new join nodes as edges and new
 // leaf nodes as vertices to the joinOrderBuilder graph, returning
 // the subgraph's newly tracked vertices and edges.
-func (j *joinOrderBuilder) populateSubgraph(n sql.Node) (vertexSet, edgeSet) {
+func (j *joinOrderBuilder) populateSubgraph(n sql.Node) (vertexSet, edgeSet, *exprGroup) {
+	var group *exprGroup
 	startV := j.allVertices()
 	startE := j.allEdges()
 	// build operator
 	switch n := n.(type) {
+	case *plan.Filter:
+		group = j.buildFilter(n)
+	case *plan.Limit:
+		group = j.buildLimit(n)
 	case *plan.JoinNode:
-		j.buildJoinOp(n)
+		group = j.buildJoinOp(n)
 	case sql.Nameable:
-		j.buildJoinLeaf(n)
+		group = j.buildJoinLeaf(n)
 	case *plan.StripRowNode:
 		return j.populateSubgraph(n.Child)
 	case *plan.CachedResults:
 		return j.populateSubgraph(n.Child)
 	default:
 		panic(fmt.Sprintf("expected Nameable node, found: %T", n))
 	}
-	return j.allVertices().difference(startV), j.allEdges().Difference(startE)
+	return j.allVertices().difference(startV), j.allEdges().Difference(startE), group
 }
 
-func (j *joinOrderBuilder) buildJoinOp(n *plan.JoinNode) {
-	leftV, leftE := j.populateSubgraph(n.Left())
-	rightV, rightE := j.populateSubgraph(n.Right())
+func (j *joinOrderBuilder) buildLimit(n *plan.Limit) *exprGroup {
+	actionableChild := n.Child
+
+	filter, isChildFilter := actionableChild.(*plan.Filter)
+	if isChildFilter {
```

Comment:
seems like we can probably just build child recursively, and then tag limit onto the group returned. if you have LIMIT->SELECT or SELECT->LIMIT the recursion works regardless. (Also SELECT -> LIMIT probably is not valid SQL?)

Reply:
Definitely, this is simpler than having the two separate functions:
```go
	case *plan.Filter:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.filter = n.Expression
	case *plan.Limit:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.limit = n.Limit
```

Reply:
Definitely simpler to have this than two extra functions:
```go
	case *plan.Filter:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.filter = n.Expression
	case *plan.Limit:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.limit = n.Limit
```